lib/util/ts_parser.js

```diff
@@ -0,0 +1,440 @@
+        const pid = ((data[start + 1] & 0x1f) << 8) + data[start + 2];
+        const atf = (data[start + 3] & 0x30) >> 4;
+
+        // if an adaption field is present, its length is specified by the
+        // fifth byte of the TS packet header.
+        let offset;
+        if (atf > 1) {
+          offset = start + 5 + data[start + 4];
+          // continue if there is only adaptation field
+          if (offset === start + 188) {
+            continue;
+          }
+        } else {
+          offset = start + 4;
+        }
+        switch (pid) {
+          case 0:
+            if (stt) {
+              offset += data[offset] + 1;
+            }
+
+            this.pmtId = this.parsePAT(data, offset);
+            break;
+          case 17:
+          case 0x1fff:
+            break;
+          case this.pmtId: {
+            if (stt) {
+              offset += data[offset] + 1;
+            }
+
+            const parsedPIDs = this.parsePMT(data, offset);
+
+            // only update track id if track PID found while parsing PMT
+            // this is to avoid resetting the PID to -1 in case
+            // track PID transiently disappears from the stream
+            // this could happen in case of transient missing audio samples
+            // for example
+            // NOTE this is only the PID of the track as found in TS,
+            // but we are not using this for MP4 track IDs.
+            if (this.videoPid == null) {
+              this.videoPid = parsedPIDs.video;
+              this.videoCodec = parsedPIDs.videoCodec;
+            }
+            if (this.audioPid == null) {
+              this.audioPid = parsedPIDs.audio;
+              this.audioCodec = parsedPIDs.audioCodec;
+            }
+            if (this.id3Pid == null) {
+              this.id3Pid = parsedPIDs.id3;
+            }
+
+            if (unknownPIDs && !this.pmtParsed) {
+              shaka.log.debug('reparse from beginning');
+              unknownPIDs = false;
+              // we set it to -188, the += 188 in the for loop will reset
+              // start to 0
+              start = syncOffset - 188;
+            }
+            this.pmtParsed = true;
+            break;
+          }
+          case this.videoPid:
+            this.videoData.push(data.subarray(offset, start + 188));
+            break;
+          case this.audioPid:
+            this.audioData.push(data.subarray(offset, start + 188));
+            break;
+          case this.id3Pid:
+            this.id3Data.push(data.subarray(offset, start + 188));
+            break;
+          default:
+            unknownPIDs = true;
+            break;
+        }
+      } else {
+        shaka.log.warning('Found TS packet that do not start with 0x47');
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Parse PAT
+   * @param {Uint8Array} data
+   * @param {number} offset
+   * @return {number}
+   */
+  parsePAT(data, offset) {
+    // skip the PSI header and parse the first PMT entry
```

Comment:
From the usage, it appears that this returns a PMT ID.  Please improve the comments explaining what you're extracting from what and why.

Reply:
Changed!